### ISSUE TITLE
Chained or clauses produces incorrect result set

### ISSUE THREAD

**Issue by TheDefinitiveMilton** [JanusGraph/janusgraph#2231]
Executing a gremlin query which contains multiple or clauses, will instead execute a single or clause combining all conditions. This yields a superset of the expected result set, as only one condition must pass. For example, two or-clauses containing two conditions each, becomes one or clause containing four clauses.

The below simple example demonstrates this issue:

```
gremlin> graph = JanusGraphFactory.build().set('storage.backend', 'inmemory').open()
==>standardjanusgraph[inmemory:[127.0.0.1]]
gremlin> mgmt = graph.openManagement()
==>org.janusgraph.graphdb.database.management.ManagementSystem.798dad3d
gremlin> sample = mgmt.makeVertexLabel('sample').make()
==>sample
gremlin> prop_a = mgmt.makePropertyKey('a').dataType(Boolean.class).make()
==>a
gremlin> prop_b = mgmt.makePropertyKey('b').dataType(Boolean.class).make()
==>b
gremlin> prop_c = mgmt.makePropertyKey('c').dataType(Boolean.class).make()
==>c
gremlin> prop_d = mgmt.makePropertyKey('d').dataType(Boolean.class).make()
==>d
gremlin> mgmt.addProperties(sample, prop_a, prop_b, prop_c, prop_d)
==>sample
gremlin> mgmt.commit()
==>null
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[inmemory:[127.0.0.1]], standard]
gremlin> g.addV('test').property('a', true).property('b', true).property('c', true).property('d', true).
......1>   addV('test').property('a', true).property('b', false).property('c', true).property('d', false).
......2>   addV('test').property('a', false).property('b', true).property('c', false).property('d', true).
......3>   addV('test').property('a', false).property('b', false).property('c', true).property('d', false)
==>v[4176]
gremlin> g.tx().commit()
==>null
gremlin> g.V().
......1>     or(
......2>         has('a', true),
......3>         has('b', true)
......4>     ).
......5>     or(
......6>         has('c', false),
......7>         has('d', true)
......8>     )
==>[a:[vp[a->true]],b:[vp[b->true]],c:[vp[c->true]],d:[vp[d->true]]]
==>[a:[vp[a->true]],b:[vp[b->false]],c:[vp[c->true]],d:[vp[d->false]]]
==>[a:[vp[a->false]],b:[vp[b->true]],c:[vp[c->false]],d:[vp[d->true]]]
```

The second vertex should not have been returned as it fails the second or clause.

**Comment by FlorianHockmann** [comment]
This indeed looks like a bug. I just confirmed that TinkerGraph shows the expected behavior:

```groovy
gremlin> TinkerGraph.open()
==>tinkergraph[vertices:0 edges:0]
gremlin> g = graph.traversal()
==>graphtraversalsource[tinkergraph[vertices:0 edges:0], standard]
gremlin> g.addV('test').property('a', true).property('b', true).property('c', true).property('d', true).
            addV('test').property('a', true).property('b', false).property('c', true).property('d', false).
            addV('test').property('a', false).property('b', true).property('c', false).property('d', true).
            addV('test').property('a', false).property('b', false).property('c', true).property('d', false)
==>v[47]
gremlin> g.V().or(has("a", true), has("b", true)).or(has("c", false), has("d", true)).valueMap()
==>[a:[true],b:[true],c:[true],d:[true]]
==>[a:[false],b:[true],c:[false],d:[true]]
```

**Comment by li-boxuan** [comment]
Seems this bug was introduced by https://github.com/JanusGraph/janusgraph/pull/927 which was released in v0.3.0.
Briefly speaking, #927 takes it for granted that "or" is only used on the top level. In this issue's example, "or" is nested under "and" condition.

A workaround would be to rewrite the query and make sure "or" is only used at the top level. For example,

(A or B) and (C or D)

can be rewritten into

(A and C) or (A and D) or (B and C) or (B and D)

which is lengthier but at least should hopefully give the correct results.

**Comment by TheDefinitiveMilton** [comment]
Unfortunately in my use case, these traversals are generated programmatically in such a way that I don't know all the clauses up front in order to attempt the workaround.

This problem is further exasperated in cases where I apply subgraph strategies such as for access control.
